homeassistant/components/intellifire/__init__.py

```diff
@@ -31,82 +42,165 @@
 ]
 
 
+def _construct_common_data(entry: ConfigEntry) -> IntelliFireCommonFireplaceData:
+    """Convert a config entry into IntelliFireCommonFireplaceData."""
+
+    return IntelliFireCommonFireplaceData(
+        auth_cookie=entry.data[CONF_AUTH_COOKIE],
+        user_id=entry.data[CONF_USER_ID],
+        web_client_id=entry.data[CONF_WEB_CLIENT_ID],
+        serial=entry.data[CONF_SERIAL],
+        api_key=entry.data[CONF_API_KEY],
+        ip_address=entry.data[CONF_IP_ADDRESS],
+        read_mode=entry.options[CONF_READ_MODE],
+        control_mode=entry.options[CONF_CONTROL_MODE],
+    )
+
+
+async def _async_pseudo_migrate_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> ConfigEntry:
+    """Update configuration entry to latest VERSION 1 format.."""
+    new = {**config_entry.data}
+    # Rename Host to IP Address
+    new[CONF_IP_ADDRESS] = new.pop("host")
+
+    username = config_entry.data[CONF_USERNAME]
+    password = config_entry.data[CONF_PASSWORD]
+
+    # Create a Cloud Interface
+    cloud_interface = IntelliFireCloudInterface()
+    await cloud_interface.login_with_credentials(username=username, password=password)
+
+    # See if we can find the fireplace first by serial and then secondly by IP.
+
+    serial = config_entry.title.replace("Fireplace ", "")
+
+    # If serial matches the hex style pattern we'll assume its good
+    valid_serial = bool(re.match(r"^[0-9A-Fa-f]{32}$", serial))
+
+    new_data = (
+        cloud_interface.user_data.get_data_for_serial(serial) if valid_serial else None
+    )
+    if not new_data:
+        new_data = cloud_interface.user_data.get_data_for_ip(new[CONF_IP_ADDRESS])
+
+    if not new_data:
+        raise ConfigEntryAuthFailed
+
+    # Find the correct fireplace
+    if new_data is not None:
+        new[CONF_API_KEY] = new_data.api_key
+        new[CONF_WEB_CLIENT_ID] = new_data.web_client_id
+        new[CONF_AUTH_COOKIE] = new_data.auth_cookie
+
+        new[CONF_IP_ADDRESS] = new_data.ip_address
+        new[CONF_SERIAL] = new_data.serial
+
+        config_entry.version = 1
+        hass.config_entries.async_update_entry(
+            config_entry,
+            data=new,
+            options={CONF_READ_MODE: "local", CONF_CONTROL_MODE: "local"},
+            unique_id=serial,
+        )
+        LOGGER.debug("Pseudo Migration %s successful", config_entry.version)
+
+    return config_entry
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up IntelliFire from a config entry."""
     LOGGER.debug("Setting up config entry: %s", entry.unique_id)
 
     if CONF_USERNAME not in entry.data:
-        LOGGER.debug("Old config entry format detected: %s", entry.unique_id)
+        LOGGER.debug("Super Old config entry format detected: %s", entry.unique_id)
         raise ConfigEntryAuthFailed
```

Comment:
Basically we kind of need to start from scratch - at this point - is this the correct error to raise?